system/smbus.c

```diff
@@ -1189,35 +1119,220 @@ static spd_info parse_spd_sdram(uint8_t smb_idx, uint8_t slot_idx)
-                            return i;
-                        }
+            vid = pci_config_read16(0, smbdev, smbfun, 0);
+            if (vid != 0xFFFF) {
+                did = pci_config_read16(0, smbdev, smbfun, 2);
+                if (did != 0xFFFF) {
+                    if (find_smb_controller(vid, did)) {
+                        return true;
                     }
                 }
             }
         }
     }
+    return false;
+}
+
+// PCI device IDs for Intel i801 SMBus controller.
+static const uint16_t intel_ich5_dids[] =
+{
+    0x2413, // 82801AA (ICH)
+    0x2423, // 82801AB (ICH)
+    0x2443, // 82801BA (ICH2)
+    0x2483, // 82801CA (ICH3)
+    0x24C3, // 82801DB (ICH4)
+    0x24D3, // 82801E (ICH5)
+    0x25A4, // 6300ESB
+    0x266A, // 82801F (ICH6)
+    0x269B, // 6310ESB/6320ESB
+    0x27DA, // 82801G (ICH7)
+    0x283E, // 82801H (ICH8)
+    0x2930, // 82801I (ICH9)
+    0x5032, // EP80579 (Tolapai)
+    0x3A30, // ICH10
+    0x3A60, // ICH10
+    0x3B30, // 5/3400 Series (PCH)
+    0x1C22, // 6 Series (PCH)
+    0x1D22, // Patsburg (PCH)
+    0x1D70, // Patsburg (PCH) IDF
+    0x1D71, // Patsburg (PCH) IDF
+    0x1D72, // Patsburg (PCH) IDF
+    0x2330, // DH89xxCC (PCH)
+    0x1E22, // Panther Point (PCH)
+    0x8C22, // Lynx Point (PCH)
+    0x9C22, // Lynx Point-LP (PCH)
+    0x1F3C, // Avoton (SOC)
+    0x8D22, // Wellsburg (PCH)
+    0x8D7D, // Wellsburg (PCH) MS
+    0x8D7E, // Wellsburg (PCH) MS
+    0x8D7F, // Wellsburg (PCH) MS
+    0x23B0, // Coleto Creek (PCH)
+    0x8CA2, // Wildcat Point (PCH)
+    0x9CA2, // Wildcat Point-LP (PCH)
+    0x0F12, // BayTrail (SOC)
+    0x2292, // Braswell (SOC)
+    0xA123, // Sunrise Point-H (PCH)
+    0x9D23, // Sunrise Point-LP (PCH)
+    0x19DF, // Denverton  (SOC)
+    0x1BC9, // Emmitsburg (PCH)
+    0xA1A3, // Lewisburg (PCH)
+    0xA223, // Lewisburg Super (PCH)
+    0xA2A3, // Kaby Lake (PCH-H)
+    0x31D4, // Gemini Lake (SOC)
+    0xA323, // Cannon Lake-H (PCH)
+    0x9DA3, // Cannon Lake-LP (PCH)
+    0x18DF, // Cedar Fork (PCH)
+    0x34A3, // Ice Lake-LP (PCH)
+    0x38A3, // Ice Lake-N (PCH)
+    0x02A3, // Comet Lake (PCH)
+    0x06A3, // Comet Lake-H (PCH)
+    0x4B23, // Elkhart Lake (PCH)
+    0xA0A3, // Tiger Lake-LP (PCH)
+    0x43A3, // Tiger Lake-H (PCH)
+    0x4DA3, // Jasper Lake (SOC)
+    0xA3A3, // Comet Lake-V (PCH)
+    0x7AA3, // Alder Lake-S (PCH)
+    0x51A3, // Alder Lake-P (PCH)
+    0x54A3, // Alder Lake-M (PCH)
+};
+
+static bool find_in_did_array(uint16_t did, const uint16_t * ids, unsigned int size) {
+    for (unsigned int i = 0; i < size; i++) {
+        if (*ids++ == did) {
+            return true;
+        }
+    }
+    return false;
+}
+
+
+static bool find_smb_controller(uint16_t vid, uint16_t did) {
+
+    switch(vid)
+    {
+        case VID_INTEL:
+        {
+            if (find_in_did_array(did, intel_ich5_dids, sizeof(intel_ich5_dids) / sizeof(intel_ich5_dids[0]))) {
```

Comment:
Would avoid passing the length of the array by doing something like that:

```
--- a/system/smbus.c
+++ b/system/smbus.c
@@ -1200,11 +1200,16 @@ static const uint16_t intel_ich5_dids[] =
     0x7AA3, // Alder Lake-S (PCH)
     0x51A3, // Alder Lake-P (PCH)
     0x54A3, // Alder Lake-M (PCH)
+    0xFFFF
 };

-static bool find_in_did_array(uint16_t did, const uint16_t * ids, unsigned int size) {
-    for (unsigned int i = 0; i < size; i++) {
-        if (*ids++ == did) {
+static bool find_in_did_array(uint16_t did, const uint16_t *ids)
+{
+    uint16_t curdid = 0;
+
+    while (curdid != 0xFFFF) {
+        curdid = *ids++;
+        if (curdid == did) {
             return true;
         }
     }
@@ -1212,13 +1217,14 @@ static bool find_in_did_array(uint16_t did, const uint16_t * ids, unsigned int s
 }


-static bool find_smb_controller(uint16_t vid, uint16_t did) {
+static bool find_smb_controller(uint16_t vid, uint16_t did)
+{

     switch(vid)
     {
         case VID_INTEL:
         {
-            if (find_in_did_array(did, intel_ich5_dids, sizeof(intel_ich5_dids) / sizeof(intel_ich5_dids[0]))) {
+            if (find_in_did_array(did, intel_ich5_dids)) {
                 return ich5_get_smb_base();
             }
```

Reply:
Indeed :)

Reply:
Don't forget this one ;)

Reply:
Indeed, I forgot about that one :)
However, I can now see that the binary happens to be smaller (simpler code) without the comparison against 0xFFFF... so let's keep this classic array size construct ?